### ISSUE TITLE
Android Problems Solved

### ISSUE THREAD

**Issue by SkySrfr** [don/BluetoothSerial#89]
I had some issues connecting to a bluetooth serial device with my Android 4.1.2 and my Android 4.4.2 phone. Turns out it was related to this: http://stackoverflow.com/questions/18657427/ioexception-read-failed-socket-might-closed-bluetooth-on-android-4-3?rq=1

I fixed it by changing run() in BluetoothSerialService.java to start with this:

```
    public void run() {
        Log.i(TAG, "BEGIN mConnectThread SocketType:" + mSocketType);
        setName("ConnectThread" + mSocketType);

        // Always cancel discovery because it will slow down a connection
        mAdapter.cancelDiscovery();

        // Make a connection to the BluetoothSocket
        try {
            // This is a blocking call and will only return on a successful connection or an exception
            Log.i(TAG,"Connecting to socket...");
            mmSocket.connect();
        } catch (IOException e) {
          Log.e(TAG, e.toString());
          //  e.printStackTrace();

          try {
            Log.i(TAG,"Trying fallback...");
            mmSocket =(BluetoothSocket) mmDevice.getClass().getMethod("createRfcommSocket", new Class[] {int.class}).invoke(mmDevice,1);
            mmSocket.connect();
            Log.i(TAG,"Connected");
          } catch (Exception e2) {
            Log.e(TAG, "Couldn't establish Bluetooth connection!");
            try {
              mmSocket.close();
            } catch (IOException e3) {
              Log.e(TAG, "unable to close() " + mSocketType + " socket during connection failure", e3);
            }
            connectionFailed();
            return;
          }
         // return;
        }
```

...

This also made it necessary to remove 'final' from mmSocket, but I'm sure there is a better way to execute the same idea.

If you are running this on Android and seeing "read failed, socket might closed or timeout, read ret: -1" in your log, this will most likely solve the problem.

Thanks for all your work on this plugin! Hopefully this can help someone down the line...

**Comment by yangyyapp** [comment]
I found that after I connected, and then send the command, which is the first Bluetooth device is not responding, only after disconnecting the Bluetooth device will execute the instruction. My Android phone is version 4.4.2, Bluetooth device is Bluetooth 2.0. I am very depressed！

**Comment by viren-vii** [comment]
The solution not working for me! @SkySrfr Please help :)

**Comment by SkySrfr** [comment]
> The solution not working for me! @SkySrfr Please help :)

Oh man, this was patched and closed 7 years ago. Assuming you are using the latest version, I would guess you are having a different problem, but I haven't looked at this stuff in years, so I doubt I can be of any help. Maybe file a new ticket?

**Comment by viren-vii** [comment]
> Oh man, this was patched and closed 7 years ago. Assuming you are using the latest version, I would guess you are having a different problem, but I haven't looked at this stuff in years, so I doubt I can be of any help. Maybe file a new ticket?

Okay! Thanks for replying :DDD

**Comment by vaibhiarora03** [comment]
@viren-vii did anything worked for you?

**Comment by viren-vii** [comment]
@vaibhiarora03 I worked on [this](https://github.com/kenjdavidson/react-native-bluetooth-classic) which worked fine with connections of Bluetooth classic devices.

**Comment by vaibhiarora03** [comment]
@viren-vii  okay. thanks.
